### DISCUSSION TITLE
Running code locally while in Docker mode

### DISCUSSION THREAD

**Discussion by minimaxir** [gitpod-io/openvscode-server#83] [Q&A] [answered]
Docker mode running on my personal iMac with the provided CLI arguments works great for running VS Code in the browser. However, it's unclear what the best practices are for running code written with VS Code locally, as all Terminal commands would be run in the sparse Docker container itself (in contrast to the chonky default container included with Codespaces).

What's the best way to work around this? Is there something similar to a `.devcontainer` as Codespaces does for installing only what you need?

**Comment by gbraad** [comment, marked as answer]
Simplest solutions are:

  * use/create a specific container image as the base for the server
  * use ssh to connect to the underlying host

Or avoid the container and use the server directly on the remote host. 

Note: this is not a replacement for codespaces, but merely the server component as used by them. Any infrastructure or deployment is not part of this repo.

**Comment by minimaxir** [comment]
Although infra may not explicitly be a part of this repo, I suspect locally hosting the server but accessing it remotely will be a common use case (for example, a Hacker News user [asked](https://news.ycombinator.com/item?id=28685213) "with docker, how would you install eg. node and go binaries?"). Since this repo has guides for Cloud deployment, there should also be guides for local deployment.

In my case, I can't `avoid the container and use the server directly on the remote host` because I need to be able to use it on my iMac/macOS as a host and the only distribution is for Linux, and `use ssh to connect to the underlying host` is high-friction, especially since I want to be able to use it on my iPad connecting to the iMac via LAN in the browser (and OpenVSCode-server works _much_ better than code-server in this case!)

Therefore, the final option is `use/create a specific container image as the base for the server` which will probably work the best, although it may require somewhat more clever Dockerfile engineering (e.g. ~downloading the large server binary as a BUILD image and COPY it to the base image w/ the relevant packages so it's cached and doesn't need to be rebuilt~ using `ADD`). I can investigate that.

**Comment by minimaxir** [comment]
So here's a Dockerfile for running basic Python, and using the Linux binary.

```Dockerfile
FROM busybox AS build-env
ENV OPENVSCODE_SERVER_VERSION 1.60.2
ADD https://github.com/gitpod-io/openvscode-server/releases/download/openvscode-server-v${OPENVSCODE_SERVER_VERSION}/openvscode-server-v${OPENVSCODE_SERVER_VERSION}-linux-x64.tar.gz /tmp
RUN cd /tmp/ && \
    tar -xzf openvscode-server-v${OPENVSCODE_SERVER_VERSION}-linux-x64.tar.gz && \
    mv openvscode-server-v${OPENVSCODE_SERVER_VERSION}-linux-x64 vscode && \
    rm openvscode-server-v${OPENVSCODE_SERVER_VERSION}-linux-x64.tar.gz

### CUSTOM CONTAINER SPECS HERE
FROM python:3-slim
### CUSTOM CONTAINER SPECS HERE

COPY --from=build-env /tmp/ /
WORKDIR /vscode
ENTRYPOINT ["./server.sh"]
```
Seems to work (you can run `python` in the terminal), and the container is actually half the size of `gitpod/openvscode-server` on disk!

Not entirely sure if this will be convenient for multiple projects, but it's easy enough to tweak, especially since the untarred files cache.
